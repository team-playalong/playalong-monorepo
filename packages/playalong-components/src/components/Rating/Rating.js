import React from 'react';
import PropTypes from 'prop-types';
import 'font-awesome/css/font-awesome.css';
import FontAwesome from 'react-fontawesome';
import styled from 'styled-components';
import COLORS from '../../utils/colors';

const RatingComp = styled.ul`
  margin: 0;
  padding: 0;
  display: inline-block;
  li {
    padding: 1px;
    color: #ddd;
    font-size: 20px;
    list-style-type: none;
    display: inline-block;
    cursor: pointer;
    &.filled {
      color: ${COLORS.STARS};
    }
    &.non-clickable {
      cursor: default;
    }
  }
`;

const Rating = ({ max = 5, readonly = false, value = null, click = null }) => {
  const classNames = `rating ${readonly ? 'readonly' : ''}`;

  function renderRatingOptions(props) {
    const options = [];
    let classes;
    for (let i = 0; i < max; i++) {
      classes = `
        star
        ${value > i ? 'filled' : ''}
        ${readonly ? 'non-clickable' : ''}
      `;
      options.push((
        <li
          className={classes}
          key={i}
          onClick={() => ratingOptionClicked(i + 1)}
        >
          <FontAwesome
            name='star'
          />
        </li>
      ));
    }
    return options;
  }

  function ratingOptionClicked(val) {
    if (!readonly && typeof click === 'function') {
      click(val);
    }
  }

  return (
    <RatingComp className={classNames}>
      {renderRatingOptions(props)}
    </RatingComp>
  );
};
Rating.propTypes = {
  readonly: PropTypes.bool,
  max: PropTypes.number,
  value: PropTypes.number,
  click: PropTypes.func,
};
export const props = ['readonly', 'max', 'value', 'click'];

export default Rating;
