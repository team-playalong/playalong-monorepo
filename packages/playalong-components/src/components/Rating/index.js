export {default, props} from './Rating';
