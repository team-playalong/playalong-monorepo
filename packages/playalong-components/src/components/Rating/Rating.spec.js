import * as React from 'react';
import { shallow } from 'enzyme';
import renderer from 'react-test-renderer';
import Rating from './Rating';

let wrapper;
const props = {
  readonly: true,
  max: 5,
  value: 3,
  click: jest.fn(),
};

function getWrapper() {
  return shallow(<Rating {...props} />);
}

beforeEach(() => {
  wrapper = getWrapper();
});

test.only('Rating should have a readonly class', () => {
  expect(wrapper.hasClass('readonly')).toBe(true);
});

test('Rating should render all rating options', () => {
  const options = wrapper.find('li');
  expect(options.length).toBe(props.max);
});

test('Rating should fill stars according to value', () => {
  const options = wrapper.find('li.filled');
  expect(options.length).toBe(props.value);
});

test('Rating should respond to click on one of the options', () => {
  props.readonly = false;
  getWrapper().find('li').first().simulate('click');
  expect(props.click.mock.calls[0][0]).toBe(1);
});

test('Snapshot testing', () => {
  const tree = renderer.create(<Rating
    readonly={true}
    max={5}
    value={3}
    click={function () {}} />).toJSON();

  expect(tree).toMatchSnapshot();
});
