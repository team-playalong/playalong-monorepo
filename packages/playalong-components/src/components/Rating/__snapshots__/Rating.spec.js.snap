// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot testing 1`] = `
<ul
  className="rating readonly sc-bdVaJa eUxWUg"
>
  <li
    className="
          star
          filled
          non-clickable
        "
    onClick={[Function]}
  >
    <span
      aria-hidden={true}
      className="fa fa-star"
    />
  </li>
  <li
    className="
          star
          filled
          non-clickable
        "
    onClick={[Function]}
  >
    <span
      aria-hidden={true}
      className="fa fa-star"
    />
  </li>
  <li
    className="
          star
          filled
          non-clickable
        "
    onClick={[Function]}
  >
    <span
      aria-hidden={true}
      className="fa fa-star"
    />
  </li>
  <li
    className="
          star
          
          non-clickable
        "
    onClick={[Function]}
  >
    <span
      aria-hidden={true}
      className="fa fa-star"
    />
  </li>
  <li
    className="
          star
          
          non-clickable
        "
    onClick={[Function]}
  >
    <span
      aria-hidden={true}
      className="fa fa-star"
    />
  </li>
</ul>
`;
